core/iwasm/common/gc/gc_object.c

```diff
@@ -712,9 +712,18 @@ wasm_object_get_ref_list(WASMObjectRef obj, bool *p_is_compact_mode,
     }
     else if (rtt_type->defined_type->type_flag == WASM_TYPE_ARRAY) {
         /* array object */
-        *p_is_compact_mode = true;
-        *p_ref_num = (uint16)wasm_array_obj_length((WASMArrayObjectRef)obj);
-        *p_ref_start_offset = (uint16)offsetof(WASMArrayObject, elem_data);
+        WASMArrayType *type = (WASMArrayType *)rtt_type->defined_type;
+        if (wasm_is_type_reftype(type->elem_type)) {
+            *p_is_compact_mode = true;
+            *p_ref_num = (uint16)wasm_array_obj_length((WASMArrayObjectRef)obj);
```

Comment:
A concern for the current design is that uint16 only supports maximum 65535 array elements, will it be too small for some scenarios? Should we extend it?

Reply:
Maybe enhance in another patch?
This patch still have some issue in fast interpreter, in `wasm_interp_traverse_gc_rootset`, some non-gc managed objects are added into gc root, seems the frame_ref not work well

Reply:
Yes, let's merge this PR first.

Reply:
updated in #2206